ui/src/services/file-explorer-service.ts

```diff
@@ -0,0 +1,116 @@
+import { makeAutoObservable } from 'mobx'
+import { inject, injectable } from 'inversify'
+
+import { FSListMessage } from '@/types/fs-list-message.type'
+
+import { CONTAINER_IDS } from '@/config/inversify/container-ids'
+import { DeviceControlStore } from '@/store/device-control-store'
+import { deviceConnectionRequired } from '@/config/inversify/decorators'
+
+import { S_IFDIR, S_IFLNK, S_IFMT } from '@/constants/file-bit-masks'
+
+@injectable()
+@deviceConnectionRequired()
+export class FileExplorerService {
+  currentPath = '/'
+  fsList: FSListMessage[] = []
+
+  constructor(@inject(CONTAINER_IDS.deviceControlStore) private deviceControlStore: DeviceControlStore) {
+    makeAutoObservable(this)
+
+    this.listDirectory()
+  }
+
+  get sortedFsList(): FSListMessage[] {
+    return this.fsList.slice().sort((prev, next) => {
+      const isPrevDirectory = this.isDirectory(prev.mode)
+      const isNextDirectory = this.isDirectory(next.mode)
+
+      if (isPrevDirectory !== isNextDirectory) {
+        return isPrevDirectory ? -1 : 1
+      }
+
+      return prev.name.localeCompare(next.name)
+    })
+  }
+
+  get isFsListEmpty(): boolean {
+    return this.fsList.length === 0
+  }
+
+  setCurrentPath(value: string): void {
+    this.currentPath = value
+
+    if (!this.currentPath.startsWith('/')) {
+      this.currentPath = `/${this.currentPath}`
+    }
+  }
+
+  enterDirectoryLocation(): void {
+    if (!this.currentPath) return
+
+    this.currentPath = this.currentPath.replace(/\/\/+/g, '/')
+
+    this.listDirectory()
+  }
+
+  addSegment(value: string): void {
+    this.currentPath = this.appendPathSegment(value)
+
+    this.listDirectory()
+  }
+
+  async getFile(file: string): Promise<void> {
+    try {
+      const path = this.appendPathSegment(file)
+
+      const fsRetrieveResult = await this.deviceControlStore.fsRetrieve(path)
+      const { content } = await fsRetrieveResult.donePromise
+
+      if (content) {
+        window.open(`${content.href}?download`, '_blank')
+      }
```

Comment:
This seems inefficient when dealing with large files. Maybe we can stream the result?